cli/src/commands/operation.rs

```diff
@@ -235,6 +254,64 @@ fn cmd_op_restore(
     Ok(())
 }
 
+fn cmd_op_prune(
+    ui: &mut Ui,
+    command: &CommandHelper,
+    args: &OperationPruneArgs,
+) -> Result<(), CommandError> {
+    let workspace_command = command.workspace_helper(ui)?;
+    let repo = workspace_command.repo();
+    let op_store = repo.op_store();
+    let op_heads_store = repo.op_heads_store();
+    let old_head_op = repo.operation().clone();
+    if let Some(unwanted_op_str) = &args.ancestors {
+        let unwanted_head_op = workspace_command.resolve_single_op(unwanted_op_str)?;
+        if old_head_op == unwanted_head_op {
+            return Err(user_error("Cannot prune the current operation"));
+        }
+        // TODO: Extract lib function that walks/rewrites operations.
+        // TODO: Stop walk at unwanted_head_op if possible.
+        let unwanted_op_ids: HashSet<OperationId> = operation::walk_ancestors(&unwanted_head_op)
+            .map_ok(|op| op.id().clone())
+            .try_collect()?;
+        let wanted_ops: Vec<Operation> = operation::walk_ancestors(&old_head_op)
+            .filter_ok(|op| !unwanted_op_ids.contains(op.id()))
+            .try_collect()?;
+        // TODO: Transaction::merge_operation() doesn't support multiple roots
+        // atm. If root op id is introduced, maybe unwanted_op_ids can be mapped
+        // to the root id?
+        let mut rewritten_op_ids: HashMap<OperationId, OperationId> = HashMap::new();
+        for old_op in wanted_ops.into_iter().rev() {
+            let mut data = old_op.store_operation().clone();
+            data.parents = data
+                .parents
+                .iter()
+                .filter(|id| !unwanted_op_ids.contains(id))
+                .map(|id| {
+                    rewritten_op_ids
+                        .get(id)
+                        .expect("parent should have been rewritten")
+                })
+                .cloned()
+                .collect();
+            let new_id = op_store.write_operation(&data)?;
+            rewritten_op_ids.insert(old_op.id().to_owned(), new_id);
+        }
+        // TODO: somehow integrate with UnpublishedOperation API?
+        let new_head_op_id = rewritten_op_ids
+            .get(old_head_op.id())
+            .expect("current op should have been rewritten");
+        op_heads_store.update_op_heads(slice::from_ref(old_head_op.id()), new_head_op_id);
```

Comment:
If another command was running concurrently, this might produce two op heads that are unrelated. If that happens, it looks like we would fail to load the repo here: https://github.com/martinvonz/jj/blob/f0f57d438e83b8d6499fcec2ba2482f0e050deab/lib/src/transaction.rs#L73. Incidentally, I mentioned on Discord just a day or two ago that maybe we should introduce a virtual "root operation" similar to our virtual root commit. That would at least remove the crash. The resulting repo would probably be in a reasonable state too. And then the user can simply abandon/prune the unwanted operations again (once we've added support for pruning operations and descendants).

(no action needed)

Reply:
> If another command was running concurrently, this might produce two op heads that are unrelated.

Yes. I noticed the crash while hacking `.jj/repo/op_heads/heads` directory, and I wasn't bothered to fix it because I remembered the virtual root operation discussion. ;)

Unrelated to this, maybe we'll need to remap op ids stored in the working copies.

Reply:
I made `jj op abandon ..op_id` leave the first operation because it was easier to implement. If a virtual root id is introduced, we can use the root id instead.